packages/integrations/cloudflare/src/server.advanced.ts

```diff
@@ -42,7 +57,32 @@ export function createExports(manifest: SSRManifest) {
 					context.waitUntil(promise);
 				},
 			});
-			let response = await app.render(request, routeData);
+
+			const cloudflareContext = {} as CloudflareContext;
+
+			// We define a custom property, so we can check the value passed to locals
+			Object.defineProperty(cloudflareContext, 'locals', {
+				enumerable: true,
+				// we should protect top-level locals from being overwritten
+				// users should just use nested properties, e.g. locals.test = "foo"
+				writable: false,
+				configurable: false,
+				value: {
+					runtime: {
+						waitUntil: (promise: Promise<any>) => {
+							context.waitUntil(promise);
+						},
+						env: env,
+						cf: request.cf,
+						caches: caches,
+					},
+				},
+			});
+
+			// We should freeze the runtime object, to make sure it&nested properties are not mutated
+			Object.freeze(cloudflareContext.locals.runtime);
```

Comment:
While it makes sense to use `Object.freeze`, it provide really **poor DX**. You'll get something like:

```
// TypeError: Cannot freeze ...
```

Which doesn't give info about *why it happened* and *how to fix it*. We should create an object with `Object.create` and throw errors using the `set()` function.

Reply:
I agree, but isn't the issue with using a `set()` function, that it will only protect the exact property _(in this case `runtime`)_ and not any nested once _(e.g. `runtime.env`)_?

Reply:
Yes. In that case you want to manually create each property:

```js
const locals = Object.create({}, {
	env: {
		enumerable: true,
		value: env,
		set() {
			throw new Error()
		}
	},
	waitUntil: {
		enumerable: true,
		value: /* value */,
		set() {
			throw new Error()
		}
	}
});
```

It's a bit lengthy, but I think it's the best solution to offer a better DX